src/lib/docker-utils.ts

```diff
@@ -225,24 +239,58 @@ export async function fetchDeltaWithProgress(
 							`Got an error when parsing delta server response for v3 delta: ${e}`,
 						);
 					}
-					id = await applyBalenaDelta(name, token, onProgress, logFn);
+					// Try to apply delta DELTA_APPLY_RETRY_COUNT times, then throw DeltaApplyError
+					let lastError: StatusError | undefined = undefined;
+					for (
```

Comment:
Actually, I thought about it a bit more. There is already retry behavior builtin on the state engine. This is adding another retry loop on top of the existing one. The counter should live somewhere in memory and be increased when a new pull is retried and revert back to regular pull when the count has reached `DELTA_APPLY_RETRY_COUNT`

That code can probably live along the `imageFetchFailures` counter we already have here https://github.com/balena-os/balena-supervisor/blob/dad6eeea73d03592eb4a665379bed6b1335b8df5/src/compose/images.ts#L172

Reply:
Since this re-try should not pull anything again but try to apply the already pulled delta, then I think it's OK to leave the retry behavior here.